game/i18n.hh

```diff
@@ -2,26 +2,130 @@
 
 #include <boost/locale.hpp>
 #include <iostream>
+#include <string>
+#include <map>
+
+#include "configuration.hh"
 #include "fs.hh"
 
 #define _(x) boost::locale::translate(x).str()
 #define translate_noop(x) x
 
 class TranslationEngine {
 public:
-	TranslationEngine(const char *package) {
-		boost::locale::generator gen;
-		gen.add_messages_path(getLocaleDir().string());
-		gen.add_messages_domain(package);
-		try {
-			std::locale::global(gen(boost::locale::util::get_system_locale(true)));
-		} catch (...) {
-			std::clog << "locale/warning: Unable to detect locale, will try to fallback to en_US.UTF-8" << std::endl;
-			std::locale::global(gen("en_US.UTF-8"));
-		}
+	TranslationEngine(const char *package) : m_package(package) {
+
+		initializeAllLanguages();
+		populateLanguages(GetAllLanguages());
+
+		std::clog << "locale/debug: Checking if language is set in config." << std::endl;
+
+		auto prefValue = config["game/language"].getEnumName();
+		auto lang = getLanguageByHumanReadableName(prefValue);
+
+		setLanguage(lang);
 	};
 	static bool enabled() {
 		return true;
 	};
 
+	static ConfigItem& translationConfig() {
+		static ConfigItem& backend = config["game/language"];
+		return backend;
+	}
+
+	void initializeAllLanguages() {
+		auto path = getLocaleDir().string();
+		m_gen.add_messages_path(path);
+		m_gen.add_messages_domain(m_package);
+		m_gen.locale_cache_enabled(true);
+
+		for (auto const& language : GetAllLanguages()) {
+			if (language.first == "None") continue;
+			m_gen(language.first);
+		}
+	}
+
+	void setLanguage(const std::string& language, bool fromSettings = false) {
+		std::clog << "locale/notice: Setting language to: '" << language << "'" << std::endl;
+
+		if (fromSettings) {
+			m_currentLanguage = { getLanguageByHumanReadableName(language) , language };
+		}
+		else {
+			m_currentLanguage = { language , getLanguageByKey(language) };
+		}
+
+		try {
+			std::locale::global(m_gen(m_currentLanguage.first));
+			std::cout << "locale/notice: Current language is: '" << m_currentLanguage.second << "'" << std::endl;
+			populateLanguages(GetAllLanguages(true), true);
+		}
+		catch (...) {
```

Comment:
is that really throwing *anything*?

Reply:
![image](https://user-images.githubusercontent.com/3918517/131903335-479bc5e2-7c93-4ccc-a7a4-f1375648f89d.png)
I guess we can catch runtime exception agreed?

Reply:
Catching runtime exception now